src/flush.rs

```diff
@@ -39,41 +43,71 @@ where S: Body,
 
     /// Try to flush the body.
     fn poll_complete(&mut self) -> Poll<(), h2::Error> {
-        let mut first = try_ready!(self.poll_body());
+        use self::DataOrTrailers::*;
 
         loop {
-            if let Some(DataOrTrailers::Data(buf)) = first {
-                let second = self.poll_body()?;
-                let eos = if let Async::Ready(None) = second {
-                    true
-                } else {
-                    false
-                };
-                self.h2.send_data(buf, eos)?;
-                if eos {
+            match try_ready!(self.poll_body()) {
+                Some(Data(buf)) => {
+                    let eos = self.body.is_end_stream();
+
+                    self.h2.send_data(SendBuf::new(buf.into_buf()), eos)?;
+
+                    if eos {
+                        self.state = FlushState::Done;
+                        return Ok(Async::Ready(()));
+                    }
+                }
+                Some(Trailers(trailers)) => {
+                    self.h2.send_trailers(trailers)?;
+                    return Ok(Async::Ready(()));
+                }
+                None => {
+                    // If this is hit, then an EOS was not reached via the other
+                    // paths. So, we must send an empty data frame with EOS.
+                    self.h2.send_data(SendBuf::none(), true)?;
+
                     return Ok(Async::Ready(()));
-                } else if let Async::Ready(item) = second {
-                    first = item;
-                } else {
-                    return Ok(Async::NotReady);
                 }
-            } else if let Some(DataOrTrailers::Trailers(trailers)) = first {
-                self.h2.send_trailers(trailers)?;
-                return Ok(Async::Ready(()));
-            } else {
-                return Ok(Async::Ready(()));
             }
         }
     }
 
     /// Get the next message to write, either a data frame or trailers.
-    fn poll_body(&mut self) -> Poll<Option<DataOrTrailers<S::Data>>, h2::Error> {
+    fn poll_body(&mut self)
+        -> Poll<Option<DataOrTrailers<S::Data>>, h2::Error>
+    {
         loop {
             match self.state {
                 FlushState::Data => {
+                    // Before trying to poll the next chunk, we have to see if
+                    // the h2 connection has capacity. We do this by requesting
+                    // a single byte (since we don't know how big the next chunk
+                    // will be.
+                    self.h2.reserve_capacity(1);
+
+                    if self.h2.capacity() == 0 {
+                        // TODO: The loop should not be needed once
+                        // carllerche/h2#270 is fixed.
+                        loop {
+                            match try_ready!(self.h2.poll_capacity()) {
+                                Some(0) => {}
```

Comment:
What prevents this from looping forever?

Reply:
That is not the behavior in `h2` 🗡 (this code is to work around an h2 bug, in theory Some(0) should never be returned).

Reply:
Though if the bug did exist, we'd then turn the CPU up to 11 and watch the place burn down...